### PR TITLE
Contract ids instead of usernames

### PR DESCRIPTION
**Pull request by corrideat**

### CODE REVIEW

File: frontend/controller/actions/group.js [hunk 1]
```diff
@@ -181,6 +181,7 @@ export default (sbp('sbp/selectors/register', {
           settings: {
             // authorizations: [contracts.CanModifyAuths.dummyAuth()], // TODO: this
             groupName: name,
+            groupCreator: userID,
```
Comment (taoeffect) [inline frontend/controller/actions/group.js]: I'm not a big fan of creating an ambiguity between usernames and identity contractIDs, and that's what this variable name does.

In various places in the app, a variable like `groupCreator` could refer to the username, or the identity contract ID. I think we should be explicit about when it's an ID instead of a username to avoid confusion.

So for this variable I would choose `groupCreatorID` or `groupCreatorIdentityContractID` instead.

File: frontend/controller/actions/group.js [hunk 2]
```diff
@@ -262,15 +263,15 @@ export default (sbp('sbp/selectors/register', {
     // JOINING_GROUP, so that we process leave actions and don't interfere
     // with the leaving process (otherwise, the side-effects will prevent
     // us from fully leaving).
-    sbp('chelonia/contract/wait', [params.originatingContractID, params.contractID])
+    await sbp('chelonia/contract/wait', [params.originatingContractID, params.contractID])
```
Comment (taoeffect) [inline frontend/controller/actions/group.js:266]: Looks like you caught a bug from a previous PR that went unnoticed?

File: frontend/controller/actions/group.js [hunk 3]
```diff
@@ -352,7 +353,7 @@ export default (sbp('sbp/selectors/register', {
         // permissions specified in the parameters.
         // The second stap is sending an OP_KEY_REQUEST message to the
         // group contract.
-        // Note that this is a two-step process that involves writing to
+        // Note  that this is a two-step process that involves writing to
```
Comment (taoeffect) [inline frontend/controller/actions/group.js]: Extra space added for no reason

File: frontend/controller/actions/group.js [hunk 4]
```diff
@@ -262,15 +263,15 @@ export default (sbp('sbp/selectors/register', {
...
     sbp('okTurtles.data/set', 'JOINING_GROUP-' + params.contractID, true)
     try {
       const { loggedIn } = sbp('state/vuex/state')
       if (!loggedIn) throw new Error('[gi.actions/group/join] Not logged in')
 
-      const { username, identityContractID: userID } = loggedIn
+      const { identityContractID: userID } = loggedIn
 
-      await sbp('chelonia/contract/sync', params.contractID)
+      await sbp('chelonia/contract/sync', params.contractID, { deferredRemove: true })
```
Comment (taoeffect) [inline frontend/controller/actions/group.js:286]: This `deferredRemoved` has been added with no explanation. Please add a comment

File: frontend/controller/actions/group.js [hunk 5]
```diff
@@ -372,6 +373,7 @@ export default (sbp('sbp/selectors/register', {
           console.error(`[gi.actions/group/join] Error while sending key request for ${params.contractID}:`, e?.message || e, e)
           throw e
         })
+        sbp('chelonia/contract/cancelRemove', params.contractID)
```
Comment (taoeffect) [inline frontend/controller/actions/group.js:392]: Please add a comment here too

File: frontend/controller/actions/group.js [hunk 6]
```diff
@@ -464,47 +465,49 @@ export default (sbp('sbp/selectors/register', {
...
         console.info('Requested to join group but already waiting for OP_KEY_SHARE. contractID=' + params.contractID)
+        sbp('chelonia/contract/cancelRemove', params.contractID)
       } else {
         console.warn('Requested to join group but the state appears invalid. contractID=' + params.contractID, { sendKeyRequest, hasSecretKeys, pendingKeyShares })
       }
     } catch (e) {
       console.error('gi.actions/group/join failed!', e)
       throw new GIErrorUIRuntimeError(L('Failed to join the group: {codeError}', { codeError: e.message }))
+    } finally {
+      await sbp('chelonia/contract/remove', params.contractID, { removeIfPending: true })
```
Comment (taoeffect) [inline frontend/controller/actions/group.js:488]: What is this, why is it here, please add a comment explaining. This seems like a significant change with no explanation.

**Review by taoeffect — CHANGES_REQUESTED** [review]
Preliminary partial review

File: frontend/controller/actions/group.js [hunk 7]
```diff
@@ -813,12 +807,12 @@ export default (sbp('sbp/selectors/register', {
             })
           } catch (e) {
             if (attempt > 3) {
-              console.error(`autoBanSenderOfMessage: max attempts reached. Error ${e.message} attempting to ban ${username}`, message, e)
+              console.error(`autoBanSenderOfMessage: max attempts reached. Error ${e.message} attempting to ban ${member}`, message, e)
             } else {
               const randDelay = randomIntFromRange(0, 1500)
-              console.warn(`autoBanSenderOfMessage: ${e.message} attempting to ban ${username}, retrying in ${randDelay} ms...`, e)
+              console.warn(`autoBanSenderOfMessage: ${e.message} attempting to ban ${member}, retrying in ${randDelay} ms...`, e)
```
Comment (taoeffect) [inline frontend/controller/actions/group.js]: We still want to print out their username here as well.

I haven't finished reviewing the PR yet so I don't know, but if there isn't a getter to go from `identityContractID` => `username` then this is a great time/place to use it and add one

File: frontend/controller/actions/identity.js [hunk 8]
```diff
@@ -616,14 +619,14 @@ export default (sbp('sbp/selectors/register', {
...
+      data: {}
     })
 
-    for (const profile of partnerProfiles) {
+    for (const partner of partners) {
       await sbp('gi.actions/chatroom/join', {
         ...omit(params, ['options', 'contractID', 'data', 'hooks']),
         contractID: message.contractID(),
-        data: { username: profile.username }
+        data: { member: partner }
```
Comment (taoeffect) [inline frontend/controller/actions/identity.js]: I think we should use `memberID` everywhere you are using `member` (as an identity contract), and here `partnerID` as well.

File: frontend/main.js [hunk 9]
```diff
@@ -149,10 +150,10 @@ async function startApp () {
           errorNotification('handleEvent', e, message)
         }
       },
-      processError: (e: Error, message: GIMessage) => {
+      processError: (e: Error, message: GIMessage, msgMeta) => {
```
Comment (taoeffect) [inline frontend/main.js]: What type is this `msgMeta`? I think we should create a type for it.

File: shared/domains/chelonia/internals.js [hunk 10]
```diff
@@ -1516,7 +1516,23 @@ export default (sbp('sbp/selectors/register', {
...
+        this.config.hooks.processError?.(e, message, {
+          signingKeyId,
+          get signingContractID () {
+            return getContractIDfromKeyId(contractID, signingKeyId, state)
+          },
+          innerSigningKeyId,
+          get innerSigningContractID () {
+            return getContractIDfromKeyId(contractID, innerSigningKeyId, state)
+          }
+        })
```
Comment (taoeffect) [inline shared/domains/chelonia/internals.js]: Maybe put this object `msgMeta` into a function that returns the object instead, instead of inlining the object, and call the function here.

File: test/backend.test.js [hunk 11]
```diff
@@ -182,6 +182,16 @@ describe('Full walkthrough', function () {
...
+        },
+        {
+          id: creator.signingKeyId(),
+          name: 'creator',
+          purpose: ['sig'],
+          ringLevel: 1,
+          permissions: '*',
+          allowedActions: '*',
+          data: sbp('chelonia/rootState')[creator.contractID()]._vm.authorizedKeys[creator.signingKeyId()].data,
+          foreignKey: `sp:${encodeURIComponent(creator.contractID())}?keyName=${encodeURIComponent('csk')}`
```
Comment (taoeffect) [inline test/backend.test.js]: What is this, why was it added?

**Review by taoeffect — CHANGES_REQUESTED** [review]
Round 2

File: frontend/model/contracts/group.js [hunk 12]
```diff
@@ -1830,9 +1805,9 @@ sbp('chelonia/defineContract', {
 
         await sbp('gi.actions/chatroom/join', {
           contractID: chatRoomId,
-          data: { username: member },
+          data: actor === member ? {} : { member },
           encryptionKeyId,
-          ...username === member && {
+          ...actor === member && {
```
Comment (taoeffect) [inline frontend/model/contracts/group.js]: `actor` => `actorID`

File: frontend/controller/actions/group.js [hunk 13]
```diff
@@ -714,19 +716,12 @@ export default (sbp('sbp/selectors/register', {
...
         ...omit(params, ['options', 'action'])
       })
     }),
-  ...encryptedAction('gi.actions/group/removeOurselves',
-    (e) => L('Failed to leave group. {codeError}', { codeError: e.message }),
-    async function (sendMessage, params) {
-      await sendMessage({
-        ...omit(params, ['options', 'action'])
-      })
-    }),
```
Comment (taoeffect) [inline frontend/controller/actions/group.js:729]: Why did you decide to remove this?

See also other comment below:

> Before, contracts could not remove someone else from another group. Now they can.

File: frontend/main.js [hunk 14]
```diff
@@ -118,8 +118,9 @@ async function startApp () {
           'chelonia/storeSecretKeys', 'chelonia/crypto/keyId',
           'chelonia/queueInvocation',
           'chelonia/contract/waitingForKeyShareTo',
+          'chelonia/contract/successfulKeySharesByContractID',
           'gi.actions/chatroom/leave',
-          'gi.actions/group/removeOurselves',
+          'gi.actions/group/removeMember',
```
Comment (taoeffect) [inline frontend/main.js]: Before, contracts could not remove someone else from another group. Now they can.

File: frontend/model/contracts/chatroom.js [hunk 15]
```diff
@@ -131,8 +125,8 @@ sbp('chelonia/defineContract', {
     chatRoomAttributes (state, getters) {
       return getters.currentChatRoomState.attributes || {}
     },
-    chatRoomUsers (state, getters) {
-      return getters.currentChatRoomState.users || {}
+    chatRoomMembers (state, getters) {
+      return getters.currentChatRoomState.members || {}
```
Comment (taoeffect) [inline frontend/model/contracts/chatroom.js:129]: `members` => `memberIDs`

File: frontend/model/contracts/chatroom.js [hunk 16]
```diff
@@ -170,55 +163,58 @@ sbp('chelonia/defineContract', {
...
+        }
         if (!state.onlyRenderMessage) {
-          if (state.users[username]) {
-            throw new Error(`Can not join the chatroom which ${username} is already part of`)
+          if (state.members[member]) {
+            throw new Error(`Can not join the chatroom which ${member} is already part of`)
           }
 
-          Vue.set(state.users, username, { joinedDate: meta.createdDate })
+          Vue.set(state.members, member, { joinedDate: meta.createdDate })
```
Comment (taoeffect) [inline frontend/model/contracts/chatroom.js]: `members` => `memberIDs`

File: frontend/model/notifications/mainNotificationsMixin.js [hunk 17]
```diff
@@ -151,16 +151,16 @@ const periodicNotificationEntries = [
                   proposalData: proposal.data.proposalData,
                   expires_date_ms: proposal.data.expires_date_ms,
                   createdDate: proposal.meta.createdDate,
-                  creator: proposal.meta.username
+                  creator: proposal.creator
```
Comment (taoeffect) [inline frontend/model/notifications/mainNotificationsMixin.js]: if these are IDs then `creator` => `creatorID`

File: frontend/model/notifications/selectors.js [hunk 18]
```diff
@@ -21,7 +21,7 @@ sbp('sbp/selectors/register', {
 
     // Creates the notification object in a single step.
     const notification = {
-      avatarUsername: template.avatarUsername || sbp('state/vuex/getters').ourUsername,
+      avatarUserID: template.avatarUserID || sbp('state/vuex/getters').ourIdentityContractId,
```
Comment (taoeffect) [inline frontend/model/notifications/selectors.js:24]: Nice job here on `avatarUsername` => `avatarUserID` 😄

File: frontend/model/state.js [hunk 19]
```diff
@@ -305,13 +303,19 @@ const getters = {
       })()
     }
   },
+  userUsername (state, getters) {
+    return (user) => {
+      const profile = getters.ourContactProfilesById[user]
+      return profile?.username || user
+    }
+  },
```
Comment (taoeffect) [inline frontend/model/state.js]: `userUsername` => `usernameFromID`

**Review by taoeffect — CHANGES_REQUESTED** [review]
Round 3!

File: frontend/controller/actions/group.js (hunk 1 above)
Comment (corrideat) [reply frontend/controller/actions/group.js]: This has now been done in general and everything that is an ID has or should have a name indicating so. An exception was made for dictionaries that are indexed by ID. So, for example, `state.profiles[userID]` is used instead of `state.profileIDs[userID]`.

File: frontend/controller/actions/group.js (hunk 2 above)
Comment (corrideat) [reply frontend/controller/actions/group.js:266]: Right. As it was it did nothing.

File: frontend/controller/actions/group.js (hunk 5 above)
Comment (corrideat) [reply frontend/controller/actions/group.js:392]: Done

File: frontend/controller/actions/group.js (hunk 4 above)
Comment (corrideat) [reply frontend/controller/actions/group.js:286]: Done.

File: frontend/controller/actions/group.js (hunk 7 above)
Comment (corrideat) [reply frontend/controller/actions/group.js]: There is (as you also found out in another later comment), it was `userUsername`, which now is `usernameFromID`.

File: frontend/main.js (hunk 9 above)
Comment (corrideat) [reply frontend/main.js]: Done

File: test/backend.test.js (hunk 11 above)
Comment (corrideat) [reply test/backend.test.js]: I added a comment. TL;DR: Since now the validate function checks for signatures as part of its permission checking, we need to add an inner (unencrypted) signature here. It is good to also see that unencrypted inner signatures work.

File: frontend/controller/actions/group.js (hunk 13 above)
Comment (corrideat) [reply frontend/controller/actions/group.js:729]: I added this back but calling `removeMember`. As for removing it from the _contract_, it was to keep things DRY.

File: frontend/model/contracts/chatroom.js (hunk 15 above)
Comment (corrideat) [reply frontend/model/contracts/chatroom.js:129]: https://github.com/okTurtles/group-income/pull/1829#discussion_r1481801163

File: frontend/model/notifications/mainNotificationsMixin.js (hunk 17 above)
Comment (corrideat) [reply frontend/model/notifications/mainNotificationsMixin.js]: https://github.com/okTurtles/group-income/pull/1829#discussion_r1481801163

File: frontend/model/contracts/chatroom.js (hunk 16 above)
Comment (corrideat) [reply frontend/model/contracts/chatroom.js]: https://github.com/okTurtles/group-income/pull/1829#discussion_r1481801163

File: frontend/model/state.js (hunk 19 above)
Comment (corrideat) [reply frontend/model/state.js]: Done

File: frontend/main.js (hunk 14 above)
Comment (corrideat) [reply frontend/main.js]: Good catch. Fixed it back.

File: frontend/model/contracts/group.js (hunk 12 above)
Comment (corrideat) [reply frontend/model/contracts/group.js]: https://github.com/okTurtles/group-income/pull/1829#discussion_r1481801163

File: frontend/controller/actions/group.js (hunk 3 above)
Comment (corrideat) [reply frontend/controller/actions/group.js]: Fixed

File: frontend/controller/actions/group.js (hunk 6 above)
Comment (corrideat) [reply frontend/controller/actions/group.js:488]: Added comment

File: frontend/controller/actions/identity.js (hunk 8 above)
Comment (corrideat) [reply frontend/controller/actions/identity.js]: https://github.com/okTurtles/group-income/pull/1829#discussion_r1481801163

File: shared/domains/chelonia/internals.js (hunk 10 above)
Comment (corrideat) [reply shared/domains/chelonia/internals.js]: Done making it a function

File: frontend/views/containers/dashboard/GroupMembersActivity.vue [hunk 20]
```diff
@@ -86,20 +86,20 @@ export default ({
...
         .sort((a, b) => b[1] - a[1])
-        .map(([username, streak]) => L('{user} - {count} month streak', { user: this.userDisplayName(username), count: streak }))
+        .map(([username, streak]) => L('{user} - {count} month streak', { user: this.userDisplayNameFromID(username), count: streak }))
     },
     missedPayments () {
       return Object.entries(this.groupStreaks.missedPayments || {})
         .filter(([username, streak]) => streak >= STREAK_MISSED_PAYMENTS)
         .map(([username, streak]) => {
-          const Largs = { user: this.userDisplayName(username), streak }
+          const Largs = { user: this.userDisplayNameFromID(username), streak }
```
Comment (taoeffect) [inline frontend/views/containers/dashboard/GroupMembersActivity.vue]: `userDisplayNameFromID` sounds like it takes an ID but here it's taking a username.. can fix?

File: frontend/views/pages/DesignSystem.vue [hunk 21]
```diff
@@ -1061,8 +1061,8 @@ page(
             td
               users-selector(
                 label='Search for users'
-                :usernames='form.searchUsers'
-                defaultValue='alexjin'
+                :users='form.searchUsers'
```
Comment (taoeffect) [inline frontend/views/pages/DesignSystem.vue]: If `:users` is supposed to take `memberIDs`, can you modify it take be specific (since "users" is unclear whether it's usernames or memberIDs)? (either `userIDs` or `memberIDs` is fine)

File: frontend/views/pages/Join.vue [hunk 22]
```diff
@@ -95,8 +95,9 @@ export default ({
   },
   mounted () {
     // For some reason in some Cypress tests it loses the route query when initialized is called
-    this.ephemeral.query = this.$route.query
-    if (syncFinished || !this.ourUsername) {
+    this.ephemeral.hash = new URLSearchParams(this.$route.hash.slice(1))
+    console.error(this.$route)
```
Comment (taoeffect) [inline frontend/views/pages/Join.vue]: Looks like some debug logging that needs to be removed

File: frontend/model/contracts/group.js [hunk 23]
```diff
@@ -1830,65 +1809,100 @@ sbp('chelonia/defineContract', {
...
       }
-      if (data.member === username && sbp('okTurtles.data/get', 'JOINING_GROUP-' + contractID)) {
-        console.info(`[gi.contracts/group/leaveGroup] for ${contractID}: member is currently joining`, { contractID, member: data.member, status: state.profiles?.[data.member]?.status })
+      if (memberID === identityContractID && sbp('okTurtles.data/get', 'JOINING_GROUP-' + contractID)) {
+        console.info(`[gi.contracts/group/leaveGroup] for ${contractID}: member is currently joining`, { contractID, memberID, status: state.profiles?.[memberID]?.status })
         return
       }
 
-      leaveAllChatRoomsUponLeaving(state, data.member, meta).catch((e) => {
+      // TODO: Use this later
```
Comment (taoeffect) [inline frontend/model/contracts/group.js]: "use this later" for what? (can elaborate so we don't forget?)

File: frontend/views/containers/chatroom/MessageBase.vue [hunk 24]
```diff
@@ -200,20 +200,22 @@ export default ({
...
       return text
-        .split(new RegExp(`(${possibleMentions.join('|')})`))
-        .map(t => possibleMentions.includes(t)
-          ? { type: TextObjectType.Mention, text: t }
-          : {
-              type: TextObjectType.Text,
-              text: this.convertTextToMarkdown ? convertToMarkdown(t) : t
-            }
-        )
+        .split(new RegExp(`(?<=\\s|^)(${allMention}|${possibleMentions.join('|')})(?=[^\\w\\d]|$)`))
```
Comment (taoeffect) [inline frontend/views/containers/chatroom/MessageBase.vue:212]: What are `(?<=\\s|^)` and `(?=[^\\w\\d]|$)`? Can you add a comment explaining this regex?

File: frontend/views/containers/chatroom/NewDirectMessageModal.vue [hunk 25]
```diff
@@ -115,19 +115,23 @@ export default ({
...
+        .filter(user => {
+          if (user === this.ourIdentityContractId) {
             return false
           }
-          const chatRoomId = this.ourGroupDirectMessageFromUsernames(username)
+          const chatRoomId = this.ourGroupDirectMessageFromUserIds(user)
           return !chatRoomId || !this.ourGroupDirectMessages[chatRoomId].visible
         })
-        .map(username => this.ourContactProfiles[username])
+        .map(user => this.ourContactProfilesById[user])
```
Comment (taoeffect) [inline frontend/views/containers/chatroom/NewDirectMessageModal.vue]: `user` => `userID`

File: frontend/views/containers/chatroom/SendArea.vue [hunk 26]
```diff
@@ -562,6 +564,19 @@ export default ({
...
 
+      /* Process mentions in the form @username => @userID */
+      const mentionStart = makeMentionFromUsername('').all[0]
+      const availableMentions = this.members.map(memberID => memberID.username)
+      msgToSend = msgToSend.replace(
+        // This regular expression matches all @username mentions that are
+        // standing alone between spaces
+        new RegExp(`(?<=\\s|^)${mentionStart}(${availableMentions.join('|')})(?=[^\\w\\d]|$)`, 'g'),
+        (_, username) => {
+          console.log()
```
Comment (taoeffect) [inline frontend/views/containers/chatroom/SendArea.vue]: Odd empty `console.log`

File: frontend/views/containers/dashboard/GroupMembersActivity.vue [hunk 27]
```diff
@@ -111,18 +111,18 @@ export default ({
 
       return Object.entries(this.groupProfiles)
         .filter(([username, profile]) => compareISOTimestamps(now, profile.lastLoggedIn) >= STREAK_NOT_LOGGED_IN_DAYS * DAYS_MILLIS)
-        .map(([username]) => this.userDisplayName(username))
+        .map(([username]) => this.userDisplayNameFromID(username))
```
Comment (taoeffect) [inline frontend/views/containers/dashboard/GroupMembersActivity.vue]: here as well (and below too)

File: frontend/views/containers/navigation/Profile.vue [hunk 28]
```diff
@@ -38,9 +38,10 @@ export default ({
   },
   computed: {
     ...mapGetters([
-      'ourUsername'
+      'ourUsername',
+      'ourIdentityContractId'
     ]),
-    userDisplayName () {
+    userDisplayNameFromID () {
```
Comment (taoeffect) [inline frontend/views/containers/navigation/Profile.vue]: Why was this renamed to `userDisplayNameFromID`? This is not a function that takes an ID... the previous name seems fine.

It's also confusing because I think `userDisplayNameFromID` is a real getter

File: frontend/views/containers/chatroom/NewDirectMessageModal.vue (hunk 25 above)
Comment (corrideat) [reply frontend/views/containers/chatroom/NewDirectMessageModal.vue]: Will fix

File: frontend/views/containers/navigation/Profile.vue (hunk 28 above)
Comment (corrideat) [reply frontend/views/containers/navigation/Profile.vue]: That's right. I think `ourDisplayName` is a better name in this case, considering the semantics and the rest of the names in this file.

File: frontend/views/containers/chatroom/MessageBase.vue (hunk 24 above)
Comment (corrideat) [reply frontend/views/containers/chatroom/MessageBase.vue:212]: For some reason GitHub doesn't show that this has been updated

**Review by taoeffect — APPROVED** [review]
Herculean job!
